### DISCUSSION TITLE
cdf-infrastructure-* project folder and <environment>-config.json" file.

### DISCUSSION THREAD

**Discussion by soumyapatra00** [aws/aws-connected-device-framework#9] [General]
I want to know where should we place the project folder and the the json file. and from which location the modules should be run like the individual module.

I was confused how to start and connect with the aws console with my account details.

currently i am running below command from my assetlibrary folder and i have my project cdf-infrastructure-demo in the below folder and JSON inside cdf-infrastructure-demo/assetlibrary/<>.JSON

./assetlibrary$ CONFIG_LOCATION="aws-connected-device-framework/source/packages/services/cdf-infrastructure-demo"; npm run start --@cdf/assetlibrary:environment=development-local

log message received:

![console](https://user-images.githubusercontent.com/66759027/132959772-3fb9d8b8-8287-49b7-a84c-daa9ed50a6cd.png)

**Comment by boardthatpowder** [comment]
Hi @soumyapatra00. The screenshot looks confusing. The top half shows entering and running `provisioning`, but the bottom half shows `assetlibrary`. Have some log lines been removed?

Regarding config, it is best practice to keep configuration outside source code. Therefore if you have this codebase checked out to `/home/seed/aws-connected-device-framework`, I would recommend creating your config directory as a sibling of that, such as `/home/seed/cdf-configuration`.

Within `cdf-configuration` a specific directory structure and naming convention needs to be followed. There should be a directory named after the service, e.g. `assetlibrary`, then within there should be a config file per environment in the format of `{env}-config.json`. Looking at your example you are running with environment `development-local`, so your directories/files should look like the following:

```
/home
    /seed
        /aws-connected-device-framework            <<< a clone of this repo
        /cdf-configuration                         <<< sibling directory to hold all config
            /assetlibrary               
                development-local-config.json
```

If you want to run assetlibrary on its own, the entire steps would be:

```
> cd /home/seed/aws-connected-device-framework/source
> rush install
> rush update
> cd packages/services/assetlibrary
> CONFIG_LOCATION="/home/seed/cdf-configuration" npm run start --@cdf/assetlibrary:environment=development-local
```

Looking at your log it says `cat /assetlibrary/development-local-config.json: No such file or directory`. That path is missing the contents of the variable `$CONFIG_LOCATION`. Looking at your start command I think that may be due to the `;` before `npm` which will cause the `CONFIG_LOCATION` you are setting to not be passed to the next command. So try either removing `;` or add `export` before `CONFIG_LOCATION`.

We have docs on [Application Configuration](https://github.com/aws/aws-connected-device-framework/blob/main/source/docs/application-configuration.md) and [Development getting started](https://github.com/aws/aws-connected-device-framework/blob/main/source/docs/development/quickstart.md). If you notice errors in any of those, something missing, or something needs to be made more clear, can you help by updating and raise a PR?  Thanks

**Comment by soumyapatra00** [comment]
Thank you. will surely update if somethings needs to be clear or missed. I appreciate for you quick reply.
I and one issue in the build too, there were 4 services where builds failed:

==[ FAILURE: 5 projects ]======================================================

--[ FAILURE: @cdf/assetlibrary ]-------------------[ 5 minutes 49.7 seconds ]--
cp: cannot create directory 'dist/types/definitions': No such file or directory
--[ FAILURE: @cdf/deployment-helper ]--------------[ 2 minutes 54.0 seconds ]--
cp: cannot create directory 'dist/config': No such file or directory
--[ FAILURE: @cdf/integration-tests ]----------------[ 1 minute 3.0 seconds ]--
cp: no such file or directory: dist/step_definitions
--[ FAILURE: @cdf/provisioning ]----------------------------[ 35.78 seconds ]--
cp: cannot create directory 'dist/things/policies': No such file or directory
--[ FAILURE: @cdf/simulation-manager ]----------------------[ 15.82 seconds ]--
cp: cannot create directory 'dist/config': No such file or directory


Projects failed to build.
**[build_error.txt](https://github.com/aws/aws-connected-device-framework/files/7148916/build_error.txt)**
STPES done:
1)curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.37.2/install.sh | bash
2)nvm use v14
3)clone the repo 
4)npm install -g @microsoft/rush
5)clone repo
6)cd pathtoworkspace/aws-connected-device-framework/source >rush install
                                                                                                         >rush update
                                                                                                         >rush build


let me know if its and issue or some steps that i have missed.

**Reply by boardthatpowder** [reply]
The last line of your `build_error.txt` is complaining about a missing rush dependency file which implies something did not go right with your `rush update`, hence why `rush build` would then fail.

Can you run the following and post the `rush-update.log`:

```sh
cd source
rush purge
rush update  --debug-package-manager > rush-update.log
```

**Reply by soumyapatra00** [reply]
attached log.
[rush_update.log](https://github.com/aws/aws-connected-device-framework/files/7150046/rush_update.log)

**Reply by boardthatpowder** [reply]
Hi @soumyapatra00. That is not the full log. Did you wait for `rush update` to finish rather than terminating it? Will take longer given the extra debug information it is writing to the log. If it runs, last line should read something like `Rush update finished successfully. (1 minute 32.8 seconds)`.

Were you able to run `rush build` after the `rush update`?

FYI, I've just cloned a fresh copy and run through the building process with no issues. May be environment related. Can you post your OS version, rush version, node version, npm version, etc.

**Reply by soumyapatra00** [reply]
Yes it was the full log. No was not able to build after the update, will try with the fresh project and update you on the same.
 
Below are the details you asked for
[09/13/21]seed@VM:/.../source$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description: Ubuntu 16.04.2 LTS
Release: 16.04
Codename: xenial
[09/13/21]seed@VM:/.../source$ rush -v
Rush Multi-Project Build Tool 5.47.0 - https://rushjs.io
Node.js version is 14.17.6 (LTS)
usage: rush [-h] [-d] ...
rush: error: too few argument
[09/13/21]seed@VM:/.../source$ rush version
Rush Multi-Project Build Tool 5.47.0 - https://rushjs.io
Node.js version is 14.17.6 (LTS)
Starting "rush version"
Validating package manager shrinkwrap file.
[09/13/21]seed@VM:/.../source$ rush version --version-policy POLICY
Rush Multi-Project Build Tool 5.47.0 - https://rushjs.io
Node.js version is 14.17.6 (LTS)
Starting "rush version"
Validating package manager shrinkwrap file.
[09/13/21]seed@VM:/.../source$
[09/13/21]seed@VM:/.../source$ npm version
{
npm: '6.14.15',
ares: '1.17.2',
brotli: '1.0.9',
cldr: '39.0',
icu: '69.1',
llhttp: '2.1.3',
modules: '83',
napi: '8',
nghttp2: '1.42.0',
node: '14.17.6',
openssl: '1.1.1l',
tz: '2021a',
unicode: '13.0',
uv: '1.41.0',
v8: '8.4.371.23-node.76',
zlib: '1.2.11'
}

**Reply by boardthatpowder** [reply]
I've just ran through all the steps on an Ubuntu 16 docker image to make sure there were no issues. All ran successfully (search in log for `root@5e2bbf35d8ed` to see all the commands I ran).
[ubuntu_16.log](https://github.com/aws/aws-connected-device-framework/files/7156119/ubuntu_16.log)

As your `rush_update.log` is only part complete, I suspect something is killing your `rush update` process locally. If it failed, I would have expected to see a failure message, but instead it just stopped. If that is not able to complete successfully then you will not be able to proceed with `rush build`.

**Reply by soumyapatra00** [reply]
I could resolve the build error.
The main error for the build error was during the rush install,there was one missing dependency file which is as below

"The shrinkwrap file contains the following issues:
  Missing dependency "lodash" (4.17.21) required by "@cdf/assetlibrary-export" "

 rush update --full command helped me to fix the problem instead of just rush update. 
 Followed below Rush git link to solve it.
https://github.com/microsoft/rushstack/issues/828
